### PR TITLE
[DPE-4249] Makes create_user error more specific

### LINKED ISSUE #277: `OpenSearchUserManager.create_user` does not distinguish between creation error and user already exists
The `_start_opensearch` constantly fails with errors below at `_post_start_init`.

```
unit-opensearch-0: 15:04:27 ERROR unit.opensearch/0.juju-log creating user monitor failed
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 826, in _start_opensearch
    self._post_start_init(event)
  File "/var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 925, in _post_start_init
    self._put_or_update_internal_user_leader(COSUser)
  File "/var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 1115, in _put_or_update_internal_user_leader
    self.user_manager.put_internal_user(user, hashed_pwd)
  File "/var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_users.py", line 309, in put_internal_user
    self.create_user(COSUser, roles, hashed_pwd)
  File "/var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_users.py", line 179, in create_user
    raise OpenSearchUserMgmtError(f"creating user {user_name} failed")
charms.opensearch.v0.opensearch_users.OpenSearchUserMgmtError: creating user monitor failed
```

That happens because the `create_user` does not distinguish between failing to create an user because it actually failed or the user already exists. The later means the `response` returns `status_code==200` and `status=="OK"`. The method `create_role` has a similar problem but a better error handling logic:

```
        if resp.get("status") != "CREATED" and not (
            resp.get("status") == "OK" and "updated" in resp.get("message")
        ):
            logging.error(f"Couldn't create role: {resp}")
            raise OpenSearchUserMgmtError(f"creating role {role_name} failed")
```

### PR DESCRIPTION
**Pull request by phvalguima**
Closes #277

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 1]
```diff
@@ -815,7 +818,17 @@ def _start_opensearch(self, event: _StartOpenSearch) -> None:  # noqa: C901
         if self.opensearch.is_started():
             try:
                 self._post_start_init(event)
-            except (OpenSearchHttpError, OpenSearchNotFullyReadyError):
+            except (
+                OpenSearchHttpError,
+                OpenSearchStartTimeoutError,
+                OpenSearchNotFullyReadyError,
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:824]: `_post_start_init` doesn't raise a `OpenSearchStartTimeoutError`
```suggestion
                OpenSearchHttpError,
                OpenSearchNotFullyReadyError,
```

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 2]
```diff
@@ -815,7 +818,17 @@ def _start_opensearch(self, event: _StartOpenSearch) -> None:  # noqa: C901
...
+            ):
+                event.defer()
+            except (OpenSearchStartError, OpenSearchUserMgmtError) as e:
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:827]: `_post_start_init` doesn't raise a `OpenSearchStartError`
```suggestion
            except OpenSearchUserMgmtError as e:
```

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -815,7 +818,17 @@ def _start_opensearch(self, event: _StartOpenSearch) -> None:  # noqa: C901
...
+                # Either generic start failure or cluster is not read to create the internal users
+                logger.exception(e)
+                self.node_lock.release()
+                self.status.set(BlockedStatus(ServiceStartError))
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:831]: i'm not sure about this one - the service effectively started, the user management failure is unrelated to the start

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Pedro - a couple of comments to clarify

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 3 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:831]: Hum, makes sense. I also want to remove that `logger.exception()` from the user error path.

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by Mehdi-Bendriss — COMMENTED** [review]
The tests are not passing

**Pull request closed without merging**